### PR TITLE
query pop

### PR DESCRIPTION
**Pull request by Anwesh1**
Hey, I'm trying to get the query to pop from the final list. Over here you can see that I set an if statement where it states 
`if (query.query.uid) in top_k_indicies:`

'query.query.uid` is `9887103` and `top_k_indicies` is `[9887103, 30049242, 22936248, 33535164]`
 (It has extra PMID's because I added them)

The question is regarding the first index because `9887103` is present in the list, so it should go into the if statement. However, this is not the case. Could you possibly tell me why this is happening?

https://github.com/PathwayCommons/semantic-search/blob/1598015e545862e99ca7b1fbcdf2f22714de6a1e/semantic_search/main.py#L117-L125

I'm printing a few things there just for debugging purposes. You will be able to that the query is actually in the list. This is the output of all the print statements:

image

### CODE REVIEW

**Comment by jvwong** [comment]
> Hey, I'm trying to get the query to pop from the final list. Over here you can see that I set an if statement where it states
> `if (query.query.uid) in top_k_indicies:`
> 
Are the `top_k_indices`  the same type as query.uid?

**Comment by Anwesh1** [comment]
I'm not fully sure because it is being converted to a list using some FAISS commands. @JohnGiorgi do you know if they types are the same or different?

**Comment by JohnGiorgi** [comment]
Yeah what @jvwong said, can you check if `type(query.query.uid) == type(top_k_indicies[-1])`, otherwise your `if` statement will always be `False`

**Comment by Anwesh1** [comment]
Yeah @jvwong you were right. They were `str` and `int` and caused the issue. Thanks for the help

**Comment by Anwesh1** [comment]
Now the query is no longer included in the final output. However, if `top_k` is equal to `index.ntotal` and the query is present in the `index`, it outputs a `-1`. We should be able to change that what you want.
 
Also, how do I run `black`? I ran `flake8` and there were no errors. When I run `black` it asks me for a path and I'm not sure which path I need to provide.

**Comment by JohnGiorgi** [comment]
> Also, how do I run `black`? I ran `flake8` and there were no errors. When I run `black` it asks me for a path and I'm not sure which path I need to provide.

Run:

```
black .
flake8 .
mypy .
pytest .
```

Before each commit to see if things work locally. I will eventually get around to wrapping these in a `make` file.

Can you see how much time this code adds to each request? Should be easy to test from within Insomnia or Postman or comparable.

File: semantic_search/main.py [hunk 1]
```diff
@@ -115,4 +115,10 @@ async def query(query: Query) -> List[Dict[str, float]]:
     top_k_indicies = top_k_indicies.reshape(-1).tolist()
     top_k_scores = top_k_scores.reshape(-1).tolist()
 
+    if int(query.query.uid) in (top_k_indicies):
+        index = top_k_indicies.index(int(query.query.uid))
+        del top_k_indicies[index], top_k_scores[index]
+    else:
+        del top_k_indicies[-1], top_k_scores[-1]
```
Comment (JohnGiorgi) [inline semantic_search/main.py:122]: This might be the source of your -1 error. Not sure you can write `del` statements on the same line like this?

Comment (Anwesh1) [reply semantic_search/main.py:122]: It happens even if I structure it separately on different lines. I think it is something else.

File: semantic_search/main.py [hunk 2]
```diff
@@ -106,7 +106,7 @@ async def query(query: Query) -> List[Dict[str, float]]:
         add_to_faiss_index(ids, embeddings, model.index)
 
     # Can't search for more items than exist in the index
-    top_k = min(model.index.ntotal, query.top_k)
+    top_k = min(model.index.ntotal, query.top_k) + 1
```
Comment (JohnGiorgi) [inline semantic_search/main.py:109]: I think it should be: `top_k = min(model.index.ntotal, query.top_k + 1)`. Doesn't make sense to request more items than are in the index, right?

Comment (Anwesh1) [reply semantic_search/main.py:109]: oh yeah, that was my fault. I should've seen that. But yeah that fixes it.

**Comment by Anwesh1** [comment]
For the tests, `black .` and `flake8 .` work fine but the other two cause these errors:
image

I'm pretty sure they are caused in other places, not in the code I changed.

**Comment by JohnGiorgi** [comment]
> For the tests, `black .` and `flake8 .` work fine but the other two cause these errors:
> image
> 
> I'm pretty sure they are caused in other places, not in the code I changed.

What version of python are you running

**Comment by Anwesh1** [comment]
I'm running python 3.8.5

**Comment by JohnGiorgi** [comment]
> I'm running python 3.8.5

Okay I have no idea then.

`def inputs() -> List[str]:` is valid syntax in Python 3.8

**Comment by Anwesh1** [comment]
I can't see where it is causing the error. Is it fixable by you by any chance?

**Comment by JohnGiorgi** [comment]
Uh we should figure out why you cant run `pytest` in your environment, that will be important for the project going forward

**Comment by Anwesh1** [comment]
Yeah, both `pytest .` and `mypy .` are causing errors.

**Comment by Anwesh1** [comment]
I'm getting a warning in the `conftest.py` file when the `import pytest` is run. 

image

**Comment by JohnGiorgi** [comment]
> I'm getting a warning in the `conftest.py` file when the `import pytest` is run.
> 
> image
> 
> This shouldn't happen, right?

No. I would through the pytest documentation and make sure you can run some of their examples. Particularly [here](https://docs.pytest.org/en/stable/getting-started.html).

**Comment by Anwesh1** [comment]
Everything on the `pytest` link works like it's supposed to. This is what running pytest looks like on the `semantic-search` stuff:

image

**Comment by JohnGiorgi** [comment]
> Everything on the `pytest` link works like it's supposed to. This is what running pytest looks like on the `semantic-search` stuff:
> 
> image

Yes, this looks as expected. The test will need to be updated. Previously it expected the query as part of the response. This PR changes that, so the test breaks. Feel free to try and update the test. If you can't, I will have to get to it sometime next week

**Comment by Anwesh1** [comment]
I fixed it by doing this:
https://github.com/PathwayCommons/semantic-search/blob/d14658f1911bba42a3ecf11131d1ee93c16d9da8/tests/test_main.py#L45

I'm not sure if this is the ideal way to do this but it works now.

**Comment by JohnGiorgi** [comment]
> I fixed it by doing this:
> 
> https://github.com/PathwayCommons/semantic-search/blob/d14658f1911bba42a3ecf11131d1ee93c16d9da8/tests/test_main.py#L45
> 
> I'm not sure if this is the ideal way to do this but it works now.

It will fail when the uid of the query and a document are the same but their text is not. You can leave t

**Comment by JohnGiorgi** [comment]
I see three issues:

1. Some examples in the readme are no longer accurate so they will need to be updated.
2. The examples in the readme show the `uid` of the response as a `str`, but its really and `int`.
3. This solution doesn't follow the ideal solution mentioned by @jvwong in #52, where the query would be included in the results when the user supplies `text` as well as `uid`.

I think we could solve all of this by more clearly defining the format of the requests and responses.

**Comment by Anwesh1** [comment]
So should we update the `README.md` on this PR and do the 3rd suggestion (the changes that Jeff had suggested) in a different PR?

**Comment by Anwesh1** [comment]
In continuation to my previous comment. I was wondering where the other changes are going to be on the `README.md`? I know we just need to remove the double quotes for the `uid` on the responses, so we accurately depict it as an int.

I'm a little unclear on the next steps.

**Comment by JohnGiorgi** [comment]
This PR will mean that a query is never returned as a result, so any examples in the readme depicting that need to be update d

**Comment by Anwesh1** [comment]
I will change the `top_k` to 2 for the examples because the query isn't being returned in the response.

**Comment by Anwesh1** [comment]
Also, the `-1` issue that was happening isn't happening anymore, which is surprising. It works as we want it to now.

**Comment by JohnGiorgi** [comment]
> Also, the `-1` issue that was happening isn't happening anymore. It works as we want it to now.

Was this the fix?

https://github.com/PathwayCommons/semantic-search/pull/57#discussion_r588716525

**Comment by Anwesh1** [comment]
Yeah @JohnGiorgi, sorry about the repetition. Yes, that comment fixed the problem.

**Comment by Anwesh1** [comment]
Should we go ahead and merge this PR?

File: README.md [hunk 3]
```diff
@@ -104,7 +100,7 @@ Once the server is running, you can make a POST request with:
     {
         "query": "9887103",
         "documents": ["9887103", "30049242", "22936248"],
```
Comment (JohnGiorgi) [inline README.md:102]: This still has a query id in the list of returned document ids

File: semantic_search/main.py [hunk 4]
```diff
@@ -115,4 +115,10 @@ async def query(query: Query) -> List[Dict[str, float]]:
     top_k_indicies = top_k_indicies.reshape(-1).tolist()
     top_k_scores = top_k_scores.reshape(-1).tolist()
 
+    if int(query.query.uid) in (top_k_indicies):
```
Comment (JohnGiorgi) [inline semantic_search/main.py:118]:
```suggestion
    if int(query.query.uid) in top_k_indicies:
```

File: README.md (hunk 3 above)
Comment (JohnGiorgi) [inline README.md:102]:
```suggestion
        "documents": [30049242, 22936248],
```

Comment (Anwesh1) [reply README.md:102]: isn't that the request? not the response? The request can be made with the query id present in the documents but the response doesn't have it. I think?

Comment (Anwesh1) [reply README.md:102]: I think that is an example of the request right? not the response?

Comment (JohnGiorgi) [reply README.md:102]: Whoops, you're right

**Pull request merged**